src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt

```diff
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
 	}
 
 	private fun setProjectVersion(project: Project, versionInfo: VersionInfo) {
-		project.version = versionInfo.toString()
+		project.version = configuration.customVersionToString.apply(versionInfo)
 	}
 
 	private fun setJarManifestVersion(project: Project, versionInfo: VersionInfo) {
 		val closure = object {
 			@Suppress("unused")
 			fun apply(jar: Jar) {
 				val attributes = jar.manifest?.attributes ?: throw Exception("The Jar task has no manifest.")
-				attributes.put("Implementation-Version", versionInfo.toString())
+				attributes.put("Implementation-Version", configuration.customVersionToString.apply(versionInfo))
 				attributes.put("Implementation-Sha", versionInfo.sha)
 				attributes.put("Specification-Version", "${versionInfo.major}.${versionInfo.minor}")
 			}
 		}
 		(project.tasks ?: throw Exception("The project has no tasks.")).withType(Jar::class.java, MethodClosure(closure, "apply"))
 	}
 
-	private fun exposeVersionInfoToUser(project: Project, versionInfo: VersionInfo) {
-		project.extensions.create("ZoltuGitVersioning", Extension::class.java, versionInfo)
-	}
+	private fun exposeConfigurationObject(project: Project, configurationChangeCallback: () -> Unit) = project.extensions.create("ZoltuGitVersioning", Configuration::class.java, configurationChangeCallback)
 
-	open class Extension(open val versionInfo: VersionInfo)
+	open class Configuration(private val configurationChangeCallback: () -> Unit) {
+		companion object {
+			@Suppress("unused")
+			const val NAME = "versioning"
+			private val defaultVersionToString = Function<VersionInfo, String> { versionInfo: VersionInfo ->
+				val suffix = if (versionInfo.tags != null && versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.tags}.${versionInfo.commitCount}"
+				} else if (versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.commitCount}"
+				} else if (versionInfo.tags != null) {
+					// simple versioning
+					"${versionInfo.commitCount}-${versionInfo.tags}"
+				} else {
+					// simple versioning
+					"${versionInfo.commitCount}"
+				}
+				"${versionInfo.major}.${versionInfo.minor}.$suffix".toString()
+			}
+			private val defaultCustomDescribeProcessor = Function<String, VersionInfo?> { null }
+		}
+		/**
+		 * Contains version information for consumption.
+		 */
+		lateinit var versionInfo: VersionInfo
+
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```

Comment:
I have some concerns here:

Instantiating a new instance of `VersionInfo` in a `build.gradle` script is very clunky because of Groovy's dynamic typing. For example, Intellij IDEA will be unable to resolve the `com.zoltu.gradle.plugin.VersionInfo` symbol and will display an error, even though the code compiles. It's possible to work around this issue by including the plugin as a dependency, but I think it would be better to ask the user for a regex instead of a factory.